src/node/internal/internal_fs_utils.ts

```diff
@@ -138,6 +138,15 @@ export function normalizePath(path: FilePath, encoding: string = 'utf8'): URL {
   } else {
     throw new ERR_INVALID_ARG_TYPE('path', ['string', 'Buffer', 'URL'], path);
   }
+  // In this case, we have a string path. Any ? or # characters in the
+  // path should be percent-encoded to ensure they are not treated as
+  // special characters in the URL.
+  path = path.replace(/[#?]/g, encodeURIComponent);
```

Comment:
Why does the URL parser handle this? Can you add a comment?

Reply:
Nit: also

```suggestion
  path = path.replaceAll(/[#?]/, encodeURIComponent);
```

Reply:
> Why does the URL parser handle this?  ...

I don't know what you're asking here.

Reply:
OK I remembered... # triggers hash state, and whenever we see ? it triggers query state.

```js
> new URL('https://cloudflare.com/this-is-path-#-?')
URL {
  href: 'https://cloudflare.com/this-is-path-#-?',
  origin: 'https://cloudflare.com',
  protocol: 'https:',
  username: '',
  password: '',
  host: 'cloudflare.com',
  hostname: 'cloudflare.com',
  port: '',
  pathname: '/this-is-path-',
  search: '',
  searchParams: URLSearchParams {},
  hash: '#-?'
}
```

Reply:
If you move this path to pathname setter, you won't have to call encodeURIComponent. For example:

```js
> const a = new URL('https://cloudflare.com')
undefined
> a.pathname = '/this-is-????'
'/this-is-????'
> a
URL {
  href: 'https://cloudflare.com/this-is-%3F%3F%3F%3F',
  origin: 'https://cloudflare.com',
  protocol: 'https:',
  username: '',
  password: '',
  host: 'cloudflare.com',
  hostname: 'cloudflare.com',
  port: '',
  pathname: '/this-is-%3F%3F%3F%3F',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

Reply:
This is going to be enough of an edge case and we would still need to have the `//` path case, that I think keeping it using the `replace(...)` function should be fine.